### PR TITLE
Resolve `f16` and `f128` `unimplemented!`/`FIXME`s

### PR DESCRIPTION
**Pull request by tgross35**
This removes the ICE codepaths for `f16` and `f128`. `rustc_apfloat` is used as a dependency for the parsing of these types, since their `FromStr` implementation will not be available in the standard library for a while.

changelog: make Clippy work correctly with `f16` and `f128`

(do not merge, contains toolchain updates for testing)

### CODE REVIEW

File: clippy_utils/src/consts.rs [hunk 1]
```diff
@@ -161,12 +167,18 @@ impl<'tcx> Hash for Constant<'tcx> {
...
+                f64::from(f).to_bits().hash(state);
+            },
             Self::F32(f) => {
                 f64::from(f).to_bits().hash(state);
             },
             Self::F64(f) => {
                 f.to_bits().hash(state);
             },
+            Self::F128(f) => {
+                f.to_bits().hash(state);
```
Comment (tgross35) [inline clippy_utils/src/consts.rs:180]: What should I do here? Functions aren't yet available to convert and use `f128` for everything instead.

**Comment by tgross35** [comment]
Marking a draft since I still need to update tests

**Comment by tgross35** [comment]
I updated the tests to add f16 and f128 everywhere it seemed relevant (i.e. everywhere where both `f32` and `f64` had tests). A lot of these need to be left as FIXMEs though since they rely on functions that are not yet implemented, or const eval casting which isn't done yet.

CI testing will continue to fail until the nightly it uses gets bumped to something from the last couple days or newer (after const arithmetic)

**Comment by flip1995** [comment]
See my comment here: https://github.com/rust-lang/rust-clippy/pull/12952#issuecomment-2175543673

I would just include those changes in the Rust PR

**Comment by tgross35** [comment]
Thank you for taking a look!

Do you know what might be happening at https://github.com/rust-lang/rust-clippy/actions/runs/9569119207/job/26380925098?pr=12950#step:5:1055? It looks like the fix is just deleting the entire file for some reason.

**Comment by flip1995** [comment]
Closing in favor of https://github.com/rust-lang/rust/pull/126636 

Thanks for moving it there!

**Pull request closed without merging**